src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -903,7 +903,15 @@ fn traits_implemented_by<'a>(
                 ty
             );
             // Fast path: if this is a primitive simple `==` will work
-            let saw_impl = impl_type == ty;
+            let saw_impl = impl_type == ty
+                || match (impl_type.kind(), ty.kind()) {
+                    (ty::Adt(def, _), ty::Adt(tydef, _)) => {
+                        debug!("adt def_id: {:?}", def.did);
+                        def.did == tydef.did
```

Comment:
The variable names are a bit unclear, and it'd be good to log both:

```suggestion
                    (ty::Adt(impl_def, _), ty::Adt(ty_def, _)) => {
                        debug!(?impl_def, ?ty_def);
                        impl_def.did == ty_def.did
```